core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -215,24 +230,58 @@ private void acquireConnectionAndExecute0(ClientRequestContext ctx, Endpoint end
         }
     }
 
-    private ProxyConfig getProxyConfig(SessionProtocol protocol, Endpoint endpoint) {
+    private void resolveProxyConfig(SessionProtocol protocol, Endpoint endpoint, ClientRequestContext ctx,
+                                    BiConsumer<@Nullable ProxyConfig, @Nullable Throwable> onComplete) {
         final ProxyConfig proxyConfig = factory.proxyConfigSelector().select(protocol, endpoint);
         requireNonNull(proxyConfig, "proxyConfig");
 
+        final ServiceRequestContext serviceCtx = ServiceRequestContext.currentOrNull();
+        final ProxiedAddresses capturedProxiedAddresses = serviceCtx == null ? null
+                                                          : serviceCtx.proxiedAddresses();
+
+        // DirectProxyConfig does not have proxyAddress as its field.
+        if (proxyConfig.proxyAddress() != null) {
+            final Future<InetSocketAddress> resolveFuture = addressResolverGroup
+                    .getResolver(ctx.eventLoop().withoutContext())
+                    .resolve(proxyConfig.proxyAddress());
+
+            resolveFuture.addListener(future -> {
+                if (future.isSuccess()) {
+                    final InetSocketAddress resolvedAddress = (InetSocketAddress) future.getNow();
+                    if (resolvedAddress != null && !resolvedAddress.isUnresolved()) {
+                        final ProxyConfig newProxyConfig = proxyConfig.withNewProxyAddress(resolvedAddress);
+                        onComplete.accept(maybeHAProxy(newProxyConfig, capturedProxiedAddresses), null);
+                    } else {
+                        logger.warn("Resolved address is invalid or unresolved: {}. " +
```

Comment:
I didn't understand this case - are you concerned that the resolution succeeds but the `InetAddress` does not exist?

Would it make sense to just align error handling with dns resolution for the normal code path (not using proxy)?